### ISSUE TITLE
Acquiring writer of already owned connection

### ISSUE THREAD

**Issue by deviator** [chalucha/mqtt-d#29]
I don't know it's problem in vibe or herre. Very simple loop: one task, one mqtt publisher, trying send messages in fix time steps
```
    auto publisherQ2 = runTask(()
    {
        while (mqtt.connected)
        {
            auto meas = ctrl.measures;
            mqtt.publish("measures", meas.serializeToPrettyJson, QoSLevel.QoS2);
        }
    });
```
Gets error:
```
Full error: core.exception.AssertError@../../../../.dub/packages/vibe-d-0.8.0/vibe-d/core/vibe/core/drivers/libevent2_tcp.d(424): Acquiring writer of already owned connection.
----------------
??:? _d_assert_msg [0x75c28e]
../../../../.dub/packages/vibe-d-0.8.0/vibe-d/core/vibe/core/drivers/libevent2_tcp.d:424 @safe void vibe.core.drivers.libevent2_tcp.Libevent2TCPConnection.acquireWriter() [0x6cae3a]
../../../../.dub/packages/vibe-d-0.8.0/vibe-d/core/vibe/core/drivers/libevent2_tcp.d:364 @safe ulong vibe.core.drivers.libevent2_tcp.Libevent2TCPConnection.write(const(ubyte[]), vibe.core.stream.IOMode) [0x6ca74d]
../../../../.dub/packages/vibe-d-0.8.0/vibe-d/core/vibe/core/stream.d:136 @safe void vibe.core.stream.OutputStream.write(const(ubyte[])) [0x6e91fa]
../../../../.dub/packages/vibe-mqtt-0.2.0-alpha.6/vibe-mqtt/source/mqttd/client.d:1114 @safe bool mqttd.client.MqttClient.send!(mqttd.messages.PubRel).send(mqttd.messages.PubRel) [0x6a71bd]
../../../../.dub/packages/vibe-mqtt-0.2.0-alpha.6/vibe-mqtt/source/mqttd/client.d:780 @safe void mqttd.client.MqttClient.onPubRec(mqttd.messages.PubRec) [0x69ca27]
../../../../.dub/packages/vibe-mqtt-0.2.0-alpha.6/vibe-mqtt/source/mqttd/client.d:952 @safe void mqttd.client.MqttClient.proccessData(const(ubyte[])) [0x69d5d2]
../../../../.dub/packages/vibe-mqtt-0.2.0-alpha.6/vibe-mqtt/source/mqttd/client.d:997 @safe void mqttd.client.MqttClient.listener() [0x69d8a0]
../../../../.dub/packages/vibe-d-0.8.0/vibe-d/core/vibe/core/core.d:620 void vibe.core.core.makeTaskFuncInfo!(@safe void delegate()).makeTaskFuncInfo(ref @safe void delegate()).callDelegate(vibe.core.core.TaskFuncInfo*) [0x6a36d7]
../../../../.dub/packages/vibe-d-0.8.0/vibe-d/core/vibe/core/core.d:1269 void vibe.core.core.CoreTask.run() [0x6b87fe]
??:? void core.thread.Fiber.run() [0x79f497]
??:? fiber_entryPoint [0x79f1fa]
??:? [0xffffffff]
Main thread exiting
```
if i sleep 100.msecs in task loop it's work several hours and throws too

**Comment by crimaniak** [comment]
I don't know if this related, but sometimes I had the same error with both mysql-lited and mysql-native under vibe.d

**Comment by deviator** [comment]
https://github.com/rejectedsoftware/vibe.d/issues/1888
Ludwig says it's problem may be related to not atomic write to socket

**Comment by tchaloupka** [comment]
Sorry for the long response, not much free time :(
Please let me know if the fix is ok as I didn't test it much.

**Comment by crimaniak** [comment]
Still, have it sometimes with 0.2.0-alpha.7 and latest vibe.d
```
core.exception.AssertError@../../.dub/packages/vibe-d-0.8.1/vibe-d/core/vibe/core/drivers/libevent2_tcp.d(414): Acquiring reader of already owned connection.
----------------
??:? _d_assert_msg [0x12018ce]
../../.dub/packages/vibe-d-0.8.1/vibe-d/core/vibe/core/drivers/libevent2_tcp.d:414 @safe void vibe.core.drivers.libevent2_tcp.Libevent2TCPConnection.checkReader() [0x1176713]
../../.dub/packages/vibe-d-0.8.1/vibe-d/core/vibe/core/drivers/libevent2_tcp.d:415 @safe void vibe.core.drivers.libevent2_tcp.Libevent2TCPConnection.acquireReader() [0x1176730]
../../.dub/packages/vibe-d-0.8.1/vibe-d/core/vibe/core/drivers/libevent2_tcp.d:230 @property @safe ulong vibe.core.drivers.libevent2_tcp.Libevent2TCPConnection.leastSize() [0x11756c0]
../../.dub/packages/vibe-mqtt-0.2.0-alpha.7/vibe-mqtt/source/mqttd/client.d:1045 @safe void mqttd.client.MqttClient.listener() [0x11450bb]
../../.dub/packages/vibe-d-0.8.1/vibe-d/core/vibe/core/core.d:620 void vibe.core.core.makeTaskFuncInfo!(void delegate() @safe).makeTaskFuncInfo(ref void delegate() @safe).callDelegate(vibe.core.core.TaskFuncInfo*) [0x10eb25f]
../../.dub/packages/vibe-d-0.8.1/vibe-d/core/vibe/core/core.d:1269 void vibe.core.core.CoreTask.run() [0x1164502]
??:? void core.thread.Fiber.run() [0x125998f]
??:? fiber_entryPoint [0x12596ee]
??:? [0xffffffff]

```

**Comment by crimaniak** [comment]
Maybe additional _con.amOwner() check in listener() will help

**Comment by tchaloupka** [comment]
I've checked the MQTT client code and it only calls this methods on the TCPConnection:
```
_con.connected // doesn't acquire reader or writer
_con.empty // acquires reader
_con.leastSize // acquires reader
_con.read // acquires reader
_con.write // acquires writer
_con.flush // acquires writer
_con.close // acquires both reader and writer
```

Could this be somehow related to the disconnection? As it is the only method which acquires reader and writer. At least in the `libevent2_tcp`.

The other thing can be acquiring of reader or writer which is already acquired by the same Task, which is not tested in assert, but as I checked, acquires are handled by `scope (exit)` and doesn't seem to call each other to cause to be reacquired.

There is one exception, again with disconnection.
See: https://github.com/vibe-d/vibe.d/blob/f5f35b6333d89717569a08eaffafc4dd8a60c8bc/core/vibe/core/drivers/libevent2_tcp.d#L191, there probably should be called reset() at scope (exit) which is not called anywhere.

Anyway, I'll add the separate reader and writer mutexes to the mentioned TCPConnection methods usage. But I don't really get where can be acquired reader twice as there is only one reader task to read from the connection. Just the disconnect can do that if I didn't overlook something.
